### DISCUSSION TITLE
how to fix normals when using

### DISCUSSION THREAD

**Discussion by xiaoyang-git** [zeux/meshoptimizer#381] [Q&A]
```
if (oct)
{
	int fu, fv;
	encodeOct(fu, fv, nx, ny, nz, bits);

	v[0] = int16_t(fu);
	v[1] = int16_t(fv);
	v[2] = int16_t(meshopt_quantizeSnorm(1.f, bits));
	v[3] = 0;
}
```

**Comment by zeux** [comment]
Can you clarify what you mean by "fix"?

**Reply by xiaoyang-git** [reply]
Sorry，I mean that How to decompress normal coordinates before rendering

**Reply by xiaoyang-git** [reply]
```
int nNormalQuantizationBits = 8;
float normalScale = float((1 << (nNormalQuantizationBits - 1)) - 1);
for (int k = 0; k < nVertexCount; k++)
{
	pNormals[3 * k] = (float)pCompressedNormals[2 * k] / normalScale;
	pNormals[3 * k + 1] = (float)pCompressedNormals[2 * k + 1] / normalScale;
	pNormals[3 * k + 2] = 1.0 - (fabs(pNormals[3 * k]) + fabs(pNormals[3 * k + 1]));
}
```
Can I decompress them like this ？
Looking forward to your reply, Thank you !

**Reply by zeux** [reply]
If you use octahedral filter then it's enough to divide the components by 2^(N-1)-1 where N is the width of the storage, eg when packing using 16-bit components it would mean dividing by 32767. This assumes that you call decodeFilterOct at runtime before passing the data to the renderer, as would happen when using glTF processing.

**Reply by xiaoyang-git** [reply]
OK, Thanks for your great work!
